### ISSUE TITLE
Configurable Response Headers

### ISSUE THREAD

**Issue by molearnik** [PrestaShop/PrestaShop#12970]
**This feature request is for implementing recommended Secure Response Headers at SW level**

PrestaShop websites should use [certain HTTP response headers](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Headers) to increase the overall security. Citation from [OWASP Secure Headers Project](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Main) _Once set, these HTTP response headers can restrict modern browsers from running into easily preventable vulnerabilities._

**Problems & My Observations:**

- One can implement secure headers at web server level. See [OWASP Secure Headers Best Practices](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Best_Practices). 
However, this approach has some drawbacks: 
  - PrestaShop can be hosted on Apache, NGinx, IIS and each server requires different configuration tweaks to accomplish the same thing. You can document recommendations for each server but this makes the installation procedure more complex.
  - When I checked a couple of PrestaShop webstores from your [showcase list](https://www.prestashop.com/en/showcase) I could not find one webstore that would implement Secure headers properly. I assume that not many merchants are sufficiently tech savvy to tweak web server configurations and PrestaShop Cloud Hosting offerings also seem to do not cover required web server configurations tweaks. 
- PrestaShop vulnerability scans give quite bad score to PrestaShop webstores because of missing Secure Headers.
- Magento implements Secure Headers at SW level and provides also GUI for controlling some headers.

**Proposed solution**
- All PrestaShop controllers incl. AJAX controllers should execute common "secure headers" injection routine that respects PrestaShop configuration (touched in the next point). Static content (CSS, JS files) does not really need to be considered and it is served from CDN in some cases anyway.
- Back Office should provide an Admin page for configuring injection of a subset of [Secure Headers](https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#tab=Headers). Why? PrestaShop admins should be able to enable secure headers selectively and eventually choose from sub-options that suite their needs or which do not disrupt functionality of some add-ons (e.g. admin could decide to block internal frames completely or restrict internal frames to the same origin if some Add-On still needs to rely on internal frames). This future configuration extension should list PrestaShop supported Secure headers and technically feasible values, provide brief explanation of being addressed risk as well as potential impact on the PrestaShop website (if certain header and value combination is enabled). 
I'd recommend to implement the following subset of Headers first:
  - HSTS (this option can be enabled only if Admin toggled "Enable SSL" and "Enable SSL on all pages" in the "Shop Parameters -> General" configuration screen). Refer to [cheatsheet](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/HTTP_Strict_Transport_Security_Cheat_Sheet.md)
  - "X-Frame-Options" 
  - "Content Security Policy (CSP) frame-ancestors" [directives](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Clickjacking_Defense_Cheat_Sheet.md)
  - "X-XSS-Protection"
  - "X-Content-Type-Options"
  - Proper "Cache controls" Headers for dynamic content in order malicious user could not use the browser history to view the cached page (These cache controls are now hardcoded in multiple files and the controls are not fully correct e.g. max-age is missing)
    - Cache-Control: no-cache, no-store, max-age=0, must-revalidate
    - Pragma: no-cache
    - Expires: 0

In addition, you could use the same mechanism for conditional injection of "protective" CSS and JavaScripts fragments e.g. [Best-for-now Legacy Browser Frame Breaking Script](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Clickjacking_Defense_Cheat_Sheet.md#best-for-now-legacy-browser-frame-breaking-script)

**Alternatives considered**
Creating PrestaShop "Secure Headers" plugin and leveraging PrestaShop Hooks for injecting secure headers. The issue is that  there is no single global hook to accomplish this goal. For instance, installing hooks for AJAX controllers would require to use PHP reflection techniques i.e. iterating through all PrestaShop PHP assemblies, iterating through all classes in the assemblies, examining each and finding if there is a need for our “intercepting” some AJAX controller. Therefore it is better to modify core classes and implement this extension at the root level of Controllers hierarchy.

**Comment by khouloudbelguith** [comment]
Hi @molearnik,

Thanks for your report.
@eternoendless, @PrestaShop/prestashop-core-developers what do you think?
Thanks!

**Comment by eternoendless** [comment]
I think it would be interesting to add these options in a new page (let's say "Security") in Configure > Advanced Parameters.

ping @PrestaShop/prestashop-product-team

**Comment by matks** [comment]
> I think it would be interesting to add these options in a new page (let's say "Security") in Configure > Advanced Parameters.
> 
> ping @PrestaShop/prestashop-product-team

And in this new page, maybe add a tab "handle my sessions data" ; see https://github.com/PrestaShop/PrestaShop/issues/20084 😉

**Comment by PierreRambaud** [comment]
> > I think it would be interesting to add these options in a new page (let's say "Security") in Configure > Advanced Parameters.
> > ping @PrestaShop/prestashop-product-team
> 
> And in this new page, maybe add a tab "handle my sessions data" ; see #20084

Yes, we will start to implement this issue when the Security page will be merged.

**Comment by wouwouwou** [comment]
> > > I think it would be interesting to add these options in a new page (let's say "Security") in Configure > Advanced Parameters.
> > > ping @PrestaShop/prestashop-product-team
> > 
> > 
> > And in this new page, maybe add a tab "handle my sessions data" ; see #20084
> 
> Yes, we will start to implement this issue when the Security page will be merged.

The security page has been merged for a while now. Is this issue on the developer roadmap for this fall or will it be 2024?

**Comment by bibips** [comment]
ping @PrestaShop/prestashop-product-team. Any update for this ? Prestashop still doesn't support CSP header.

**Comment by apohutenumera** [comment]
Hi, any update on this topic ?

**Comment by boo-code** [comment]
The precondition stated in 2021 - "we will start to implement this issue when the Security page will be merged" - is met. The page eternoendless described exists on `9.2.x`, at Configure > Advanced Parameters > Security: `install-dev/data/xml/tab.xml` declares `AdminParentSecurity` with `id_parent="Advanced_Parameters"`, the route `admin_security` maps to `Configure\AdvancedParameters\SecurityController::indexAction`, and `index.html.twig` renders two forms - `GeneralType`, which carries a single switch today (back office token protection, saved through `admin_security_general_save`), and `PasswordPolicyType`. The session sub-tabs matks asked for in 2020 are there too, as `AdminSecuritySessionEmployee` and `AdminSecuritySessionCustomer`.

So the part that has been waiting is not the page. What is missing is the configuration surface and, more awkwardly, the emission point.

There is no global response-header mechanism in core. The only place a header of this kind is set is `SecuredFileReaderController`, which sets `X-Content-Type-Options` on its own response - a per-controller choice, not a shop-wide one.

Emission is two paths, not one, and that is worth deciding before any form is designed. `index.php` boots `FrontKernel` and, when it throws `NotFoundHttpException`, falls through to `Dispatcher::getInstance()->dispatch()` with `FRONT_LEGACY_CONTEXT` defined as true. That legacy dispatch never produces a Symfony `Response`, so a `kernel.response` listener registered in `PrestaShopBundle` would cover the back office, the Admin API and the Symfony front-office routes, but not the legacy front controllers - which are most of the shop. Covering both means the listener plus a legacy emission point, with one shared source of configuration so the two cannot drift.

The shape that fits the existing code: a "Response headers" section added to `GeneralType` on the page that already exists, backed by an `AbstractMultistoreConfiguration` so the values are per-shop like the rest of that page, read by a `kernel.response` listener and by the legacy path.

Not built, and the reason is the one this issue has carried since 2019 rather than the page: the header set and, more importantly, each header's default. Several of these change how an existing shop renders once switched on, so shipping them enabled by default is a decision about breaking working installations, and shipping them all disabled by default is a decision about whether the feature achieves anything without merchant action. That is the specification being asked for, and it is a product call rather than a technical one.
